src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala

```diff
@@ -1704,6 +1944,504 @@ trait ChoosesMode {
+                  tourMode -> tourMode
+                    .allowedBeamModesGivenAvailableVehicles(availableVehicles, firstOrLastLeg = false)
+                    .intersect(modesToQuery)
+                )
+                .toMap
+            val firstAndLastTripModeToTourMode = BeamTourMode.values
+              .map(tourMode =>
+                tourMode -> tourMode
+                  .allowedBeamModesGivenAvailableVehicles(availableVehicles, firstOrLastLeg = true)
+                  .intersect(modesToQuery)
+              )
+              .toMap
+            val tourModeUtils = tourModeChoiceCalculator.tourExpectedMaxUtility(
+              tourModeCosts,
+              modeChoiceCalculator.modeChoiceLogit,
+              modeToTourMode,
+              Some(firstAndLastTripModeToTourMode)
+            )
+            val out = tourModeChoiceCalculator(tourModeUtils)
+
+            // We need to keep track of the chosen vehicle Id in PersonData so that we can release it and
+            // potentially give up on our tour mode choice if a route can't be constructed
+            val chosenTourVehicle: Map[EmbodiedBeamTrip, Option[Id[BeamVehicle]]] = out match {
+              case Some(tourMode) =>
+                firstLegItineraries
+                  .collect {
+                    case itin
+                        if tourMode
+                          .allowedBeamModesGivenAvailableVehicles(availableVehicles, firstOrLastLeg = true)
+                          .contains(itin.tripClassifier) =>
+                      itin.vehiclesInTrip
+                        .find(availableVehicles.map(_.id).contains)
+                        .map(vid => itin -> Some(vid))
+                  }
+                  .flatten
+                  .toMap
+              case _ => Map.empty[EmbodiedBeamTrip, Option[Id[BeamVehicle]]]
+            }
+
+            val tourModeChoiceEvent = new TourModeChoiceEvent(
+              departTime.toDouble,
+              this.id,
+              out.map(_.value).getOrElse(""),
+              currentTour,
+              availableVehicles,
+              firstAndLastTripModeToTourMode,
+              tourModeUtils,
+              modesToQuery,
+              currentActivity(choosesModeData.personData)
+            )
+            eventsManager.processEvent(tourModeChoiceEvent)
+            (out, chosenTourVehicle)
+          case Some(tripMode) =>
+            // If trip mode is already set, determine tour mode from that and available vehicles (sticking
+            // with walk based tour if the only available vehicles are shared)
+            val chosenTourModeAndVehicle =
+              getTourModeAndVehicle(
+                firstLegItineraries.filter(_.tripClassifier == tripMode),
+                availableVehicles,
+                choosesModeData.personData.currentTourPersonalVehicle
+              )
+            val (chosenTourMode, chosenTourVehicleMap) = chosenTourModeAndVehicle.headOption match {
```

Comment:
We are using headOption on a map here. Maps usually have random order. This means, we are returning a random value. If different behaviour is needed, some sorting might be needed.